gflownet/envs/crystals/composition.py

```diff
@@ -215,12 +228,21 @@ def get_mask_invalid_actions_forward(self, state=None, done=None):
             min_atoms_per_unrequired_element = 0
             max_atoms_per_unrequired_element = 0
 
+        # Check if current state is compatible with space group
```

Comment:
Ok, I've been thinking about this over the week end and I wonder if this is the right approach.

In this PR, if we detect that there is some incompatibility between composition and space group, we deactivate the EOS action. We force the GFN to continue adding atoms. 

The problem I see with that is that if we have an incompatibility, we will never be able to fix it by using only the other actions in the action space. We can't fix an issue with an incompatible number of copper ions by adding oxygen or iron. The only way that we could fix it would be to change the number of copper ions but, to prevent cycle, we don't allow the GFN to change the number of atoms for an atom type already in the composition.

So it seems that the consequence that this change will have is that, as soon as the GFN takes an action that causes an incompatiblity with the space group, it will be forced to keep adding atoms until it hit the max number of atoms/elements even if we know that the sample will never be valid.

I think that it would actually be better if, when we detect that there is an incompatiblity, we for the EOS action to avoid wasting computational resources on a doomed trajectory.

Reply:
I think that, for the space group checks, the ideal thing would be to focus on masking actions that lead to incompatible compositions rather then observe after the fact that we are in an invalid state.

Naïvely, we could use the compatibility check to validate any number of atoms that we could add between `min_atom_i` and `max_atom_i`. That would work but be about `max_atom_i - min_atom_i + 1` times more costly than the current implementation.

But we can significantly reduce that amount by looking at the Wyckoff positions a bit ourselves to determine instances in which the space group checks are superfluous. Ex : if, for a space group, the WP with the lowest multiplicity is 8, then there's never any point in doing checks for adding 1-7 atoms. 
I'm currently trying stuff on a branch to see how efficient we could make this and still have pretty good guarantees of not painting ourselves in a corner.

Reply:
But, yeah, I can definitely see cases where we won't be able to fully integrate all constraints in our action planning and we might end up with no way out of a state.
But we should still be able to reduce that number of cases quite a bit. The trickier cases will be those with very few (or no) non-fixed WP between `min_atom_i` and `max_atom_i`.

Reply:
> I think that, for the space group checks, the ideal thing would be to focus on masking actions that lead to incompatible compositions rather then observe after the fact that we are in an invalid state.

I agree, but this may be 1) involved to code and 2) potentially demanding computationally indeed. For the latter we could always store info, so I would not not do it because of that. 

> I'm currently trying stuff on a branch

Ah cool, thanks a lot!

> For me that's an argument for keeping composition selection first [...]

Well, we had sort of agreed that we would make it flexible no?

Reply:
> Well, we had sort of agreed that we would make it flexible no?

Flexible in what way? Perhaps I wasn't present then.

Reply:
@carriepl argued that it is probably better to first set the space group, then the composition, because otherwise GFlowNet may have trouble to sample certain space groups. Correct? And I argued that ideally they should be set in any order.

Reply:
@alexhernandezgarcia Exactly.